### PR TITLE
Microsoft/users/aujensen/vs2019 vs task

### PR DESCRIPTION
**Pull request by austin-jensen**
Added support for msbuild task and vsbuild task for visual studio 2019

### CODE REVIEW

File: Tasks/Common/MSBuildHelpers/PathFunctions.ps1 [hunk 1]
```diff
@@ -227,6 +237,63 @@ function Get-VisualStudio_15_0 {
...
+                                Write-Verbose $_
+                                $null = $output.AppendLine($_)
+                            }
+                        }
+                    $script:visualStudioCache['16.0'] = (ConvertFrom-Json -InputObject $output.ToString()) |
+                        Select-Object -First 1
+                }
+            } catch {
+                Write-Verbose ($_ | Out-String)
+                $script:visualStudioCache['6.0'] = $null
```
Comment (TingluoHuang) [inline Tasks/Common/MSBuildHelpers/PathFunctions.ps1]: 16.0 instead of 6.0

**Comment by TingluoHuang** [comment]
@austin-jensen  where is your task.json change that consume this common library change?

File: Tasks/Common/MSBuildHelpers/PathFunctions.ps1 [hunk 2]
```diff
@@ -227,6 +237,63 @@ function Get-VisualStudio_15_0 {
     }
 }
 
+function Get-VisualStudio_16_0 {
```
Comment (ericsciple) [inline Tasks/Common/MSBuildHelpers/PathFunctions.ps1]: consider parameterize existing function

File: Tasks/VSBuildV1/VSBuild.ps1 [hunk 3]
```diff
@@ -49,6 +49,7 @@ try {
     $msBuildVersion = $null;
     switch ("$vsVersion") {
         '' { $msBuildVersion = '14.0' ; break } # VS wasn't found. Attempt to find MSBuild 14.0 or lower.
+        '16.0' { $msBuildVersion = '16.0' ; break }
```
Comment (ericsciple) [inline Tasks/VSBuildV1/VSBuild.ps1]: make sure to bump the version for any tasks that consume the msbuildhelpers module

**Comment by jimmylewis** [comment]
The VSTestV2 task also needs to be updated to detect/support VS2019.  Will that be added to this PR, or in a separate one?

**Comment by austin-jensen** [comment]
@jimmylewis That will be done separately, as a lot of people are currently complaining about not being able to build with visual studio 2019. So we are trying to get the msbuild and vsbuild tasks out quickly

File: Tasks/Common/MSBuildHelpers/PathFunctions.ps1 [hunk 4]
```diff
@@ -17,9 +17,19 @@ function Get-MSBuildPath {
         # Only attempt to find Microsoft.Build.Utilities.Core.dll from a VS 15 Willow install
         # when "15.0" or latest is specified. In 15.0, the method GetPathToBuildToolsFile(...)
         # has regressed. When it is called for a version that is not found, the latest version
-        # found is returned instead.
+        # found is returned instead. Same for "16.0"
         [System.Reflection.Assembly]$msUtilities = $null
-        if (($Version -eq "15.0" -or !$Version) -and # !$Version indicates "latest"
+        if (($Version -eq "16.0" -or !$Version) -and # !$Version indicates "latest"
+            ($visualStudio16 = Get-VisualStudio_16_0) -and
```
Comment (TingluoHuang) [inline Tasks/Common/MSBuildHelpers/PathFunctions.ps1]: `Get-VisualStudio_16_0`?

File: Tasks/Common/MSBuildHelpers/PathFunctions.ps1 [hunk 5]
```diff
@@ -17,9 +17,19 @@ function Get-MSBuildPath {
...
+            $visualStudio16.installationPath) {
+
+            $msbuildUtilitiesPath = [System.IO.Path]::Combine($visualStudio16.installationPath, "MSBuild\Current\Bin\Microsoft.Build.Utilities.Core.dll")
+            if (Test-Path -LiteralPath $msbuildUtilitiesPath -PathType Leaf) {
+                Write-Verbose "Loading $msbuildUtilitiesPath"
+                $msUtilities = [System.Reflection.Assembly]::LoadFrom($msbuildUtilitiesPath)
+            }
+        }
+        elseif (($Version -eq "15.0" -or !$Version) -and # !$Version indicates "latest"
             ($visualStudio15 = Get-VisualStudio_15_0) -and
```
Comment (TingluoHuang) [inline Tasks/Common/MSBuildHelpers/PathFunctions.ps1]: `Get-VisualStudio_15_0`?

File: Tasks/Common/MSBuildHelpers/PathFunctions.ps1 [hunk 6]
```diff
@@ -193,16 +206,16 @@ function Get-VisualStudio_15_0 {
                             $null = $output.AppendLine($_)
                         }
                     }
-                $script:visualStudioCache['15.0'] = (ConvertFrom-Json -InputObject $output.ToString()) |
+                $script:visualStudioCache["$MajorVersion.0"] = (ConvertFrom-Json -InputObject $output.ToString()) |
```
Comment (TingluoHuang) [inline Tasks/Common/MSBuildHelpers/PathFunctions.ps1]: not sure whether you can just use `visualStudioCache[$MajorVersion]` as key, do we have to do `"$MajorVersion.0"`?

File: Tasks/MSBuildV1/task.json [hunk 7]
```diff
@@ -11,7 +11,7 @@
     "author": "Microsoft Corporation",
     "version": {
         "Major": 1,
-        "Minor": 120,
+        "Minor": 121,
```
Comment (TingluoHuang) [inline Tasks/MSBuildV1/task.json]: i think it should be the milestone number  1.146.0

**Comment by mika76** [comment]
Any news here? I'm trying to build with 2019 build tools on a TFS 2015 on site installer...

**Pull request merged**